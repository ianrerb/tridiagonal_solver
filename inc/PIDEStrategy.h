#ifndef PIDESTRATEGY_H
#define PIDESTRATEGY_H
#include "NumericStrategy.h"
#include <cmath>

class PIDEStrategy : public NumericStrategy { 
  public:
    //constructor
    PIDEStrategy(double r_=0, double q_=0, double sigma_=0, double nu_=0, double theta_=0, double y_=0):NumericStrategy(),r(r_),q(q_),sigma(sigma_),nu(nu_),theta(theta_),y(y_){ };

    //interface implementation
    double Execute(Option&) const;

    //strategy initialization
    PIDEStrategy& Initialize(const double, const double, const double, unsigned int, unsigned int);
   
    //accessor functions
    double RiskFreeRate() const {  return r; };
    double DividendRate() const {  return q; };
    double Sigma() const {  return sigma; };
    double Nu() const {  return nu; };
    double Theta() const {  return theta; };
    double Y() const {  return y; };
    
    PIDEStrategy& RiskFreeRate(const double x) { r=x; return *this; };
    PIDEStrategy& DividendRate(const double x) { q=x; return *this; };
    PIDEStrategy& Sigma(const double x) { sigma=x; return *this; };
    PIDEStrategy& Nu(const double x) { nu=x; return *this; };
    PIDEStrategy& Theta(const double x) { theta=x; return *this; };
    PIDEStrategy& Y(const double x) { y=x; return *this; };


  private:
    double r, q, sigma, nu, theta, y;
    double spot, margin, expiry;    
    double lambda_p, lambda_n;    
    
    //helper functions for computing derivatives prices
    double ComputeLambda(char) const;

    double ComputeBoundaryPoint(const unsigned int, const unsigned int) const;
    double ComputeInteriorPoint(const unsigned int, const unsigned int) const;
};

#endif
