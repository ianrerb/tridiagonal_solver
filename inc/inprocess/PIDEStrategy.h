#ifndef PIDESTRATEGY_H
#define PIDESTRATEGY_H
#include "NumericStrategy.h"

class PIDEStrategy : public NumericStrategy { 
  public:
    PIDEStrategy(double r_=0, double q_=0, double t_=0, double sigma_=0, double nu_=0, double theta_=0, double y_=0):r(r_),q(q_),t(t_),sigma(sigma_),nu(nu_),theta(theta_),y(y_){ };

    double RiskFreeRate() const {  return r; };
    double DividendRate() const {  return q; };
    double T() const {  return t; };
    double Sigma() const {  return sigma; };
    double Nu() const {  return nu; };
    double Theta() const {  return theta; };
    double Y() const {  return y; };
    
    NumericStrategy& RiskFreeRate(const double x) { r=x; return *this; };
    NumericStrategy& DividendRate(const double x) { q=x; return *this; };
    NumericStrategy& T(const double x) { t=x; return *this; };
    NumericStrategy& Sigma(const double x) { sigma=x; return *this; };
    NumericStrategy& Nu(const double x) { nu=x; return *this; };
    NumericStrategy& Theta(const double x) { theta=x; return *this; };
    NumericStrategy& Y(const double x) { y=x; return *this; };


  private:
    double r, q, t, sigma, nu, theta, y;
};

#endif
