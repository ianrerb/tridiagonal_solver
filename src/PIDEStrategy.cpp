#include"../inc/PIDEStrategy.h"

PIDEStrategy& PIDEStrategy::Initialize(const double spot_, const double margin_, const double expiry_, unsigned int M_, unsigned int N_){ 
  NewGrid(M_,N_); 
  spot = spot_;
  margin=margin_;
  expiry = expiry_;
  SetSpaceDomain((1.0-margin)*log(spot),(1.0+margin)*log(spot));
  SetTimeDomain(0,expiry);
  lambda_p = ComputeLambda('p');
  lambda_n = ComputeLambda('n');

  return *this; 
}

double PIDEStrategy::Execute(Option& TheOption) const { 
  
  //fill in tau = 0 boundary conditions
  for(unsigned int i = 0; i!=_TheGrid->Rows(); ++i)
    (*_TheGrid)(i,0)=std::max(TheOption.Strike()-exp(SpaceDomain.first+static_cast<double>(i)*DeltaX),0.0);

  for(unsigned int i = 0; i!= _TheGrid->Rows(); ++i){ 
    for(unsigned int j = 1; j!= _TheGrid->Columns(); ++j){ 
      (*_TheGrid)(i,j) = (i==0 || (i ==_TheGrid->Rows()-1)) ? ComputeBoundaryPoint(i,j) : ComputeInteriorPoint(i,j);  
    }  
  }
  PrintGrid();
  return -5; 
}

double PIDEStrategy::ComputeLambda(char toggle) const { 
  double retval;
  retval = pow(theta/(sigma*sigma),2)+2/(pow(sigma,2)*nu);
  switch(toggle){ 
  case 'p':
    retval = pow(retval,.5) - theta/(sigma*sigma);
    break;
  
  case 'n':
    retval = pow(retval,.5) + theta/(sigma*sigma);
    break; 
  } 
  return retval;
}


//not implemented yet
double PIDEStrategy::ComputeInteriorPoint(const unsigned int i,const unsigned int j) const { 
 return 0; 
}

double PIDEStrategy::ComputeBoundaryPoint(const unsigned int i, const unsigned int j) const { 
  return 1;  
}
