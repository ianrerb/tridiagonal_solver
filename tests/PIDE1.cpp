#include"../inc/PIDEStrategy.h"

using namespace std;

int main(){
  double Spot = 1300;
  EuropeanPut TheOption(1300.0,.5);
  PIDEStrategy TheStrategy(0.0025,.015,.3,.4,-.3,0);
  TheStrategy.NewGrid(20,10);
  TheStrategy.PrintGrid();
  return 0;
}
