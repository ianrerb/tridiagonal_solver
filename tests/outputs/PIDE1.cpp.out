1083.5 1 1 1 1 1 1 1 1 1 
1038.54 0 0 0 0 0 0 0 0 0 
984.246 0 0 0 0 0 0 0 0 0 
918.675 0 0 0 0 0 0 0 0 0 
839.488 0 0 0 0 0 0 0 0 0 
743.856 0 0 0 0 0 0 0 0 0 
628.366 0 0 0 0 0 0 0 0 0 
488.892 0 0 0 0 0 0 0 0 0 
320.455 0 0 0 0 0 0 0 0 0 
117.039 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 
0 1 1 1 1 1 1 1 1 1 
